### DISCUSSION TITLE
How can I dynamically set a query string variable based on a dynamic property?

### DISCUSSION THREAD

**Discussion by rappasoft** [livewire/livewire#4300] [1. Help]
This one is pretty obscure, but I don't think the use case is too far-fetched. I have [this package](https://github.com/rappasoft/laravel-livewire-tables) which I'm working on V2 from the ground up.

I have a constant issue from the community to have the need to have multiple instances of the component on the same page, but have their state be unique in the query string.

For example, you have two instances of the same datatable showing different data, but they each have a `Name` column.

The query string might look like:

`sorts[name]=desc`

But once you sort that same column on the other table on the same page, the state gets replaced in the query string.

So here was my idea, that I can not get to work.

A property called `$tableName` on the component that defaults to `table`.

Then, since this is instance based, not component based, I *can not* just have them set it on the component class itself because it would just be the same every time. So they have to set it from the call in the HTML:

```html
<livewire:users-table :user="$user1" tableName="users1" />
<livewire:users-table :user="$user2" tableName="users2" />
```

Now the problem here is I need to make a property on the component with that name, before the query string gets instantiated.

For example, it can't be in `boot` or `booted` because `$tableName` is still `table`. I can set the property in `mount` (though I don't want to because I want that available to the user to use) like so:

```php
// $tableName is `user1` or `user2` in the mount method

$this->{$this->tableName} = [];

$this->{$this->tableName} = [
    'sorts' => $this->{$this->tableName}['sorts'] ?? []
];

// I now have a dynamic method on the component unique to this instance of this component $this->user1 or $this->user2
```

This would set the property correctly, however the query string gets called before this and the value of `$tableName` is still `table`, so even if I called the `queryString` method to dynamically add the new property, the value is still the `pre user defined` value:

```php
public function queryString(): array
{
    // $tableName is still 'table' and 'user1' or 'user2'

    return [
        $this->tableName => ['except' => null],
    ];

    // So a non-existant property called $table is being tracked in the query string instead of the existing $user1 or $user2.
}
```

So I'm in a constant state of back and forth within the lifecycle hooks to get the correct data in the correct places, and I can not figure it out for the life of me.

It's just a back and forth issue of either `undefined property $table` or `undefined property $user1/$user2`.

It seems like a catch 22.

Sorry if this is confusing, but any help is appreciated.

**Comment by jonneroelofs** [comment]
Played around with it for a little bit. This seems to work for me:

```php
protected function queryString()
    {
        return [
            'tableName' => ['except' => null]
        ];
    }
```

I think with how you wrote it the result of your queryString would actually be something like:
```php
['table' => ['except' => null]]
['users1' => ['except' => null]]
['users2' => ['except' => null]]
```

But you don't have any of those properties on your component, the property you have is 'tableName'

**Reply by rappasoft** [reply]
Right, I'm trying to get them to be dynamic so they don't have to be specified since they will be different with each instance of the component. I've managed to make it work with multiple instances of the same component that are separate classes, but not multiple instances of the same component that are the same class with different initialization parameters.

**Comment by jonneroelofs** [comment]
I think this might work:

```php 
<livewire:users-table :user="$user1"  :table="['prefix' => 'users1']" />
<livewire:users-table :user="$user2" :table="['prefix' => 'users2']" />
```

And now you can add 'table' hardcoded to the queryString, but the contents of that variable are dynamic.
Maybe you could then add all the stuff that you needed in the queryString to that array? So you add 'sorts' to the $table variable?

**Reply by rappasoft** [reply]
I actually did try that, However, since it is the same root key in the query string when one table updated it ended up clearing out the portion for the other table since it couldn't tell the difference and the tables have no idea the other one exists there too.

Thanks for trying, I figured it's something I would figure out eventually but everything feels too hacky right now.

**Comment by joshhanley** [comment]
@rappasoft hmm I reckon we can come up with a solution for that, but to start with, can you show me what an ideal querystring would look like if you have your two users tables `users1` and `users2`, both with sorts on the name column?

**Reply by rappasoft** [reply]
Thanks @joshhanley. Here is what my current implementation looks like:

1. It works for multiple components on the same page that extend the table by defining a table name and array to hold the data (trying to figure out how to avoid the latter)
2. It does not work with the same extension of the table component on the same page because there's still no way to differentiate.

Example:

UserTable1.php

```php
class UsersTable1 extends DataTableComponent
{
    public $model = User::class;

    public string $tableName = 'users1';
    public array $users1 = [];

    ...
}
```

UserTable2.php

```php
class UsersTable2 extends DataTableComponent
{
    public $model = User::class;

    public string $tableName = 'users2';
    public array $users2 = [];

    ...
}
```

And the queryString method in DataTableComponent:

```php
public function queryString(): array
{
    if ($this->queryStringIsEnabled()) {
        return [
            $this->getTableName() => ['except' => null],
        ];
    }

    return [];
}

// Table name defaults to `table` if not specified (page with only one datatable component)
```

Now whenever I need to update the query string, I have a main key I can reference:

I.e.:

```php
public function setSort(string $field, string $direction): string
{
    return $this->{$this->getTableName()}['sorts'][$field] = $direction; // Hence the need to define the array with the same name
}
```

So If I have both components on the same page:

```html
<livewire:users-table1 />
<livewire:users-table2 />
```

Since they are two different components, if I sort both name columns, the query string looks like this:

```
http://tables.test?users1[sorts][name]=desc&users2[sorts][name]=asc
```

And the component reads the keys assigned and updates accordingly, what doesn't work is this:

```html
<livewire:users-table1 :user="$user" />
<livewire:users-table1 :user="$user" />
```

Since they have the same `tableName`, the URLs will be overridden.

I honestly don't even have a good idea what it would look like or what would make the most sense, my initial thoughts are this isn't a big enough problem to warrant the amount of time it would take to figure it out and muddy up the Livewire codebase to implement.

The workaround option is I have a `disableQueryString`. Which in scenario 2 would just make each table work independently via regular Livewire sending in the `$user` on mount. There would just be no state saved in the query string if you wanted to send it to someone to re-state the tables on page load.

Curious to hear your thoughts.

**Comment by joshhanley** [comment]
@rappasoft ok yeah that makes sense.

So couple of things with your original idea:

One issue I can see is dynamic class properties are being deprecated in PHP so I probably wouldn't make use of the if you are rebuilding, see here https://wiki.php.net/rfc/deprecate_dynamic_properties

In terms of how to manage the query string, there is currently an open PR #3679 for custom hydration/ dehydration of query strings, which I think is exactly what would suit your need.

Can you have a look at that PR and see if it will work for this scenario?

That way you can have all of your data as properties on the component, and also have your `$tableName` property, and when dehydrating the query string, you can construct the query string array with the table names as needed.

I think passing the table name into each instance of the table sounds like a good idea though, sort of like how we `wire:key` elements, developers would need to name each table, even if they are of the same component type.

If you want to get multiple tables of the same component running automatically without needing to pass a table name in, you could probably create a static count or something, that gets incremented in each mount method, which then is used to generate table names like `table1`, `table2`, etc.

One thing to note if you go the static count method, is you need to ensure you clear that data at the end of a request so that way it is supported by Laravel Octane. To do this we actually have an internal `flush-state` event you can hook into which is trigged by Octane at the end of the request, so you can then reset count.

Hope this helps!

**Reply by rappasoft** [reply]
Hmm, I'm not sure if that PR would fully do what I need, but I do like that in general and think it's useful. I didn't know about the dynamic property deprecation, that's a bummer and I'll have to find some other way when the time comes.

I appreciate you putting the thought into it, I'll keep chugging along and come back to it in the future, and maybe there will be other options available.

Thanks!

**Reply by joshhanley** [reply]
Ah ok, why what wouldn't it do?

No worries 🙂

**Comment by luanfreitasdev** [comment]
Hi @joshhanley  and @rappasoft , 

From what I understood in the RFC, this has nothing to do with the example above:

  _“Dynamic property” here refers to a property that has not been declared in the class. It has no relation to the access style (e.g. $user->{'na' . 'me'} is still an access to a declared property)._

Because it's a declared property, even though it's dynamic, right? I know it's another matter, but the approach is interesting. 

Thanks

**Reply by joshhanley** [reply]
@luanfreitasdev you are correct about the access style, there's nothing changing about that.

But Anthony was hoping to use dynamic properties which are generated based on the passed in table name.

So if you had this

```php
public $tablename = "users1";
```

And then tried to access a dynamic property like this

```php
$this->{$this->tablename};
```

It means it would be looking for a property `$this->users1` but that property isn't declared on the class, hence why it is dynamic. And that is what that RFC will be deprecating.

Hope this helps!

**Comment by devsquad-luan-freitas** [comment]
Thanks @joshhanley

**Reply by joshhanley** [reply]
No worries 🙂

**Comment by Lupennat** [comment]
@rappasoft
you can use aliases to handle the query string based on the table name.
However, there are 2 bugs that create incorrect behavior using arrays in the query string.
#4475 (fixed on master)
#4473

```php
public function queryString() :array
{
    return [
        'filters' => ['except' => null, 'as' => ($this->tableName ?? '' ) . 'filters'],
        'sorts' => ['except' => null, 'as' => ($this->tableName ?? '' ) . 'sorts'],
    ];
}
```

**Comment by inmanturbo** [comment]
@Lupennat unfortunately this doesn't seem to work as expected in this case. Any ideas?
https://github.com/rappasoft/laravel-livewire-tables/pull/776#issuecomment-1128920689

**Reply by luanfreitasdev** [reply]
I believe this can solve most array problems, allowing you to define an alias for each filter type. 

https://github.com/livewire/livewire/pull/4913

**Comment by moisish** [comment]
@rappasoft I am facing the same issue here, did you find a solution?
